components/eam/src/chemistry/mozart/mo_sethet.F90

```diff
@@ -573,311 +372,224 @@ subroutine sethet( het_rates, press, zmid,  phis, tfld, &
-             het_rates(i,k, clono2_ndx) = work3(i)
-          end if
-          if( brono2_ndx > 0 ) then
-             het_rates(i,k, brono2_ndx) = work3(i)
-          end if
-          if( hcl_ndx > 0 ) then
-             het_rates(i,k, hcl_ndx) = work3(i)
-          end if
-          if( n2o5_ndx > 0 ) then
-             het_rates(i,k, n2o5_ndx) = work3(i)
-          end if
-          if( hocl_ndx > 0 ) then
-             het_rates(i,k, hocl_ndx) = work3(i)
-          end if
-          if( hobr_ndx > 0 ) then
-             het_rates(i,k, hobr_ndx) = work3(i)
-          end if
-          if( hbr_ndx > 0 ) then
-             het_rates(i,k, hbr_ndx) = work3(i)
-          end if
-
-          if( soa_ndx > 0 ) then
-             het_rates(i,k,soa_ndx) = tmp1_rates(i)
-          end if
-          if( oc2_ndx > 0 ) then
-             het_rates(i,k,oc2_ndx) = tmp1_rates(i)
-          end if
-          if( cb2_ndx > 0 ) then
-             het_rates(i,k,cb2_ndx) = tmp1_rates(i)
-          end if
-          if( so4_ndx > 0 ) then
-             het_rates(i,k,so4_ndx) = tmp1_rates(i)
-          end if
-          if( sa1_ndx > 0 ) then
-             het_rates(i,k,sa1_ndx) = tmp1_rates(i)
-          end if
-          if( sa2_ndx > 0 ) then
-             het_rates(i,k,sa2_ndx) = tmp1_rates(i)
-          end if
-          if( sa3_ndx > 0 ) then
-             het_rates(i,k,sa3_ndx) = tmp1_rates(i)
-          end if
-          if( sa4_ndx > 0 ) then
-             het_rates(i,k,sa4_ndx) = tmp1_rates(i)
-          end if
 
           if( h2so4_ndx > 0 ) then
-             het_rates(i,k,h2so4_ndx) = tmp0_rates(i)
-          end if
-          if( nh4_ndx > 0 ) then
-             het_rates(i,k,nh4_ndx) = tmp0_rates(i)
-          end if
-          if( nh4no3_ndx > 0 ) then
-             het_rates(i,k,nh4no3_ndx ) = tmp0_rates(i)
-          end if
-          if( nh3_ndx > 0 ) then
-             het_rates(i,k,nh3_ndx) = max( rain(i,k) / (h2o_mol*(work1(i) + 1._r8/(xhen_nh3(i,k)*work2(i)))),0._r8 )
-          end if
-
-          if( ch3cooh_ndx > 0 ) then
-             het_rates(i,k,ch3cooh_ndx) = max( rain(i,k) / (h2o_mol*(work1(i) + 1._r8/(xhen_ch3cooh(i,k)*work2(i)))),0._r8 )
-          end if
-          if( hcooh_ndx > 0 ) then
-             het_rates(i,k,hcooh_ndx) = max( rain(i,k) / (h2o_mol*(work1(i) + 1._r8/(xhen_ch3cooh(i,k)*work2(i)))),0._r8 )
-          endif
-          if ( hcn_ndx > 0 ) then
-             het_rates(i,k,hcn_ndx     ) = max( rain(i,k) / (h2o_mol*(work1(i) + 1._r8/(xhen_hcn(i,k)*work2(i)))),0._r8 )
-          endif
-          if ( ch3cn_ndx > 0 ) then
-             het_rates(i,k,ch3cn_ndx   ) = max( rain(i,k) / (h2o_mol*(work1(i) + 1._r8/(xhen_ch3cn(i,k)*work2(i)))),0._r8 )
+             work3(icol) = satf_hno3 *  max( rain(icol,kk) / &
+                           (h2o_mol*(work1(icol) + 1._r8/(xhen_hno3(icol,kk)*work2(icol)))), 0._r8 )
+             het_rates(icol,kk,h2so4_ndx) =  work3(icol) + tmp_hetrates(icol,kk,1)
```

Comment:
Calculation of work3 and het_rates look exactly the same for different species, can we factor it out?

Reply:
yes, refactored it out.